e2e/libs/keywords/deployment_keywords.py

```diff
@@ -0,0 +1,78 @@
+from utility.constant import KIND_PERSISTENTVOLUMECLAIM
+from utility.constant import KIND_POD
+
+from workload.deployment import create_deployment
+from workload.deployment import delete_deployment
+from workload.persistentvolumeclaim import create_persistentvolumeclaim
+from workload.persistentvolumeclaim import delete_persistentvolumeclaim
+from workload.workload import check_pod_data_checksum
+from workload.workload import get_workload_pod_names
+from workload.workload import keep_writing_pod_data
+from workload.workload import write_pod_random_data
+
+
+class deployment_keywords:
+
+    def __init__(self):
+        pass
+
+    def cleanup_deployments(self):
+        cached = ResourceCache.get_kind(KIND_DEPLOYMENT)
+        if cached is None:
+            return
+
+        uncache = []
+        for name, links in cached.items():
+            delete_deployment(name)
+            uncache.append(name)
+
+            if KIND_PERSISTENTVOLUMECLAIM in links:
+                pvc_name = links[KIND_PERSISTENTVOLUMECLAIM]
+                delete_persistentvolumeclaim(pvc_name)
+                ResourceCache.delete(KIND_PERSISTENTVOLUMECLAIM, pvc_name)
+
+        for name in uncache:
+            ResourceCache.delete(KIND_DEPLOYMENT, name)
+        assert len(ResourceCache.get_kind(KIND_DEPLOYMENT)) == 0
+
+    def create_deployment(self, name, volume_type="rwo", option=""):
+        create_persistentvolumeclaim(name, volume_type, option)
+        ResourceCache.new(KIND_PERSISTENTVOLUMECLAIM, name)
+
+        create_deployment(name)
+        ResourceCache.new(KIND_DEPLOYMENT, name)
+        ResourceCache.create_or_update_link(
+            KIND_DEPLOYMENT, name, KIND_PERSISTENTVOLUMECLAIM, name
+        )
+
+        pod_name = get_workload_pod_names(name)[0]
+        ResourceCache.new(KIND_POD, pod_name)
+        ResourceCache.create_or_update_link(
+            KIND_DEPLOYMENT, name, KIND_POD, pod_name
+        )
+
+    def check_deployment_pod_data_checksum(self, deployment_name):
+        pod_name = ResourceCache.get_link(KIND_DEPLOYMENT, deployment_name, KIND_POD)
+        checksum = ResourceCache.get_link(KIND_POD, pod_name, CACHE_CHECKSUM)
+        check_pod_data_checksum(pod_name, checksum)
+
+    def get_deployment_pod_name(self, deployment_name):
+        pod_name = get_workload_pod_names(deployment_name)[0]
+        ResourceCache.create_or_update_link(
+            KIND_DEPLOYMENT, deployment_name, KIND_POD, pod_name
+        )
+
+    def keep_writing_deployment_pod_data(self, deployment_name):
+        pod_name = ResourceCache.get_link(KIND_DEPLOYMENT, deployment_name, KIND_POD)
+        return keep_writing_pod_data(pod_name)
+
+    def write_deployment_pod_random_data(self, deployment_name, size_in_mb):
+        pod_name = ResourceCache.get_link(KIND_DEPLOYMENT, deployment_name, KIND_POD)
+        data_checksum = write_pod_random_data(pod_name, size_in_mb)
+        ResourceCache.create_or_update_link(
```

Comment:
I think the implementation flow could look like:
1. Write data: Write data to the volume/workload, and record the data checksum in the volume/pod resource cache link.
1. Take snapshot 1: Take the snapshot, then create a new snapshot resource cache, and record the data checksum from the volume/pod resource cache link.
1. Write data: Continue data write to the volume/workload, and update the data checksum in the volume/pod resource cache link.
1. Take snapshot 2: Similar to the previous step, create another snapshot, and record the data checksum from the volume/pod resource cache link in the new snapshot resource cache.
1. Write data: Continue data writing. 
1. Take snapshot 3: repeat the snapshot flow.
1. Check data: verify the volume data integrity.
1. Revert to snapshot 2: restore snapshot.
1. Check data matches snapshot 2: Compare the data checksum to the checksum in the snapshot resource cach link.
1. Revert to snapshot 1: restore snapshot.
1. Check data matches snapshot 1: repeat the verification process.

Reply:
This discussion seems to be going towards integration test case migration. Let's explore this in more depth once we start working on individual cases.

Reply:
It's better to figure out and prototype this at early stage to test this design otherwise contributors have no idea how to do that (especially our team members, since they will start joining negative test cases implementation after this PR merged) and if they are implemented in a non-ideal way, more efforts will be put in to refactoring.

Reply:
> This discussion seems to be going towards integration test case migration.

It's trying to test this design's capability to deal with real-world test scenarios.